contracts/0.8.9/oracle/BaseOracle.sol

```diff
@@ -0,0 +1,342 @@
+    /// an oracle looking at the same reference slot would calculate a different hash.
+    ///
+    function getConsensusVersion() external view returns (uint256) {
+        return CONSENSUS_VERSION_POSITION.getStorageUint256();
+    }
+
+    /// @notice Sets the consensus version expected by the oracle contract.
+    ///
+    function setConsensusVersion(uint256 version) external onlyRole(MANAGE_CONSENSUS_VERSION_ROLE) {
+        _setConsensusVersion(version);
+    }
+
+    ///
+    /// Data provider interface
+    ///
+
+    /// @notice Returns the last consensus report hash and metadata.
+    ///
+    function getConsensusReport() external view returns (
+        bytes32 hash,
+        uint256 refSlot,
+        uint256 receptionTime,
+        uint256 deadlineTime,
+        bool processingStarted
+    ) {
+        ConsensusReport memory report = _storageConsensusReport().value;
+        uint256 processingRefSlot = LAST_PROCESSING_REF_SLOT_POSITION.getStorageUint256();
+        return (
+            report.hash,
+            report.refSlot,
+            report.receptionTime,
+            report.deadlineTime,
+            report.refSlot != 0 && report.refSlot == processingRefSlot
+        );
+    }
+
+    ///
+    /// Consensus contract interface
+    ///
+
+    /// @notice Called by HashConsensus contract to push a consensus report for processing.
+    ///
+    /// Note that submitting the report doesn't require the oracle to start processing it
+    /// right away, this can happen later. Until the processing is started, HashConsensus is
+    /// free to reach consensus on another report for the same reporting frame and submit it
+    /// using this same function.
+    ///
+    function submitReport(bytes32 reportHash, uint256 refSlot, uint256 deadline) external {
+        if (_msgSender() != CONSENSUS_CONTRACT_POSITION.getStorageAddress()) {
+            revert OnlyConsensusContractCanSubmitReport();
+        }
+
+        uint256 prevSubmittedRefSlot = _storageConsensusReport().value.refSlot;
+        if (refSlot < prevSubmittedRefSlot) {
+            revert RefSlotCannotDecrease(refSlot, prevSubmittedRefSlot);
+        }
+
+        uint256 prevProcessingRefSlot = LAST_PROCESSING_REF_SLOT_POSITION.getStorageUint256();
+        if (refSlot <= prevProcessingRefSlot) {
+            revert RefSlotMustBeGreaterThanProcessingOne(refSlot, prevProcessingRefSlot);
+        }
+
+        if (prevProcessingRefSlot != prevSubmittedRefSlot) {
+            emit WarnProcessingMissed(prevSubmittedRefSlot);
+        }
+
+        emit ReportSubmitted(refSlot, reportHash, deadline);
+
+        ConsensusReport memory report = ConsensusReport({
+            hash: reportHash,
+            refSlot: refSlot.toUint64(),
+            receptionTime: uint64(_getTime()),
+            deadlineTime: deadline.toUint64()
+        });
+
+        _storageConsensusReport().value = report;
+        _handleConsensusReport(report, prevSubmittedRefSlot, prevProcessingRefSlot);
```

Comment:
`prevSubmittedRefSlot` - do we need to initialize this variable? on first `submitReport` the value is 0